src/cursor/cur_version.c

```diff
@@ -0,0 +1,275 @@
+/*-
+ * Copyright (c) 2014-present MongoDB, Inc.
+ * Copyright (c) 2008-2014 WiredTiger, Inc.
+ *	All rights reserved.
+ *
+ * See the file LICENSE for redistribution information.
+ */
+
+#include "wt_internal.h"
+
+/*
+ * __curversion_set_key --
+ *     WT_CURSOR->set_key implementation for version cursors.
+ */
+static void
+__curversion_set_key(WT_CURSOR *cursor, ...)
+{
+    WT_CURSOR *table_cursor;
+    WT_CURSOR_VERSION *version_cursor;
+
+    version_cursor = (WT_CURSOR_VERSION *)cursor;
+    table_cursor = version_cursor->table_cursor;
+    table_cursor->set_key(table_cursor);
+}
+
+/*
+ * __curversion_get_key --
+ *     WT_CURSOR->get_key implementation for version cursors.
+ */
+static int
+__curversion_get_key(WT_CURSOR *cursor, ...)
+{
+    WT_CURSOR *table_cursor;
+    WT_CURSOR_VERSION *version_cursor;
+    WT_DECL_RET;
+    va_list ap;
+
+    version_cursor = (WT_CURSOR_VERSION *)cursor;
+    table_cursor = version_cursor->table_cursor;
+    va_start(ap, cursor);
+    WT_ERR(__wt_cursor_get_keyv(cursor, cursor->flags, ap));
+    WT_ERR(__wt_cursor_get_keyv(table_cursor, table_cursor->flags, ap));
+
+err:
+    va_end(ap);
+    return (ret);
+}
+
+/*
+ * __curversion_next --
+ *     WT_CURSOR->next method for version cursors.
+ */
+static int
+__curversion_next(WT_CURSOR *cursor)
+{
+    WT_CURSOR *hs_cursor;
+    WT_CURSOR_VERSION *version_cursor;
+    WT_DECL_RET;
+    WT_SESSION_IMPL *session;
+    WT_UPDATE *upd;
+
+    version_cursor = (WT_CURSOR_VERSION *)cursor;
+    hs_cursor = version_cursor->hs_cursor;
+    CURSOR_API_CALL(cursor, session, next, NULL);
+        
+    /* 
+     * - If there are untraversed updates in the update chain, return
+     * the next one. 
+     * - If there are no updates or all the updates have been exhausted,
+     * return the ondisk value.
+     * - Once the ondisk value is exhausted, check if there are any values
+     * in the history store.
+     */
+    upd = version_cursor->next_upd;
+    if (upd != NULL && !F_ISSET(version_cursor, WT_VERSION_CUR_UPDATE_EXHAUSTED)) {
+        upd = upd->next;
+        
+        if (upd == NULL)
+            F_SET(version_cursor, WT_VERSION_CUR_UPDATE_EXHAUSTED);
+    } else if (!F_ISSET(version_cursor, WT_VERSION_CUR_ON_DISK_EXHAUSTED)) {
+        F_SET(version_cursor, WT_VERSION_CUR_ON_DISK_EXHAUSTED);
+    } else if (!F_ISSET(version_cursor, WT_VERSION_CUR_HS_EXAUSTED)) {
+        /* Use the history store cursor to position on the key. */
+        hs_cursor->next(hs_cursor);
+
+        if (ret != 0)
```

Comment:
I think we need to check WT_NOTFOUND here

Reply:
Do you mean to check ret against WT_NOTFOUND rather than 0?

Reply:
yes